src/ast.h

```diff
@@ -0,0 +1,299 @@
+    piece *elif_p;
+    piece *else_p;
+} if_expr;
+
+typedef struct {
+    char *name_i;
+    expr *for_e;
+    piece *for_p;
+} for_expr;
+
+typedef struct {
+    expr *while_e;
+    piece *while_p;
+} while_expr;
+
+typedef struct {
+    expr *match_e;
+    list *match_bs; 
+} match_expr;
+
+typedef enum {
+    fn_lit_type,
+    list_lit_type,
+    record_lit_type,
+    int_lit_type,
+    float_lit_type,
+    str_lit_type,
+    rstr_lit_type,
+    true_type,
+    false_type,
+    nil_lit_type,
+} lit_type;
+
+typedef struct {
+    int64_t i;
+} int_lit;
+
+typedef struct {
+    double f;
+} float_lit;
+
+typedef struct {
+    char *s;
+} str_lit;
+
+typedef struct {
+    char *s;
+} rstr_lit;
+
+typedef struct {
+    list *fn_param_l;
+    piece *fn_p;
+} fn_lit;
+
+typedef struct {
+    expr *list_e;
+    list *list_exprs;
+} list_lit;
+
+typedef struct {
+    char *name;
+    expr *experssion;
+} record_field;
+
+typedef struct {
+    record_field *record_f;
+    list *record_fields;
+} record_lit;
+
+typedef struct {
+    lit_type type;
+    union {
+        fn_lit *fn_l;
+        list *list_l;
+        record_lit *record_l;
+        int_lit *i_val;
+        float_lit *f_val;
+        str_lit *s_val;
+        rstr_lit *rs_val;
+    } obj;
+} lit_expr;
+
+struct expr {
+    expr_type type;
+    union {
+        lit_expr *lit_e;
+        prefix_expr* pre_e;
+        infix_expr *inf_e;
+        index_expr *index_e;
+        access_expr* access_e;
+        group_expr *group_e;
+        call_expr *call_e;
+        if_expr *if_e;
+        for_expr *for_e;
+        while_expr *while_e;
+        match_expr *match_e;
+        char *ident_n;
+    } obj;
+};
+
+typedef enum {
+    expr_body_type,
+    piece_body_type,
+} match_body_type;
+
+typedef struct {
+    match_body_type type;
+    char* patt;
+    union {
+        expr *expr;
+        piece *piece;
+    } obj;
+} match_body;
+
+typedef enum {
+    int_patt_type,
+    float_patt_type,
+    str_patt_type,
+    rstr_patt_type,
+    true_patt_type,
+    false_patt_type,
+    nil_patt_type,
+    ident_patt_type,
+    list_patt_type,
+    record_patt_type,
+} pattern_type;
+
+typedef struct {
+    pattern *outer_p;
+    pattern *inner_p;
+} list_pattern_scope;
+
+typedef struct {
+    pattern *p;
+    list *patts;
+} list_patt_bruckt;
+
+typedef enum {
+    patt_brucket_type,
+    patt_scope_type,
+} list_pattern_type;
+
+typedef struct {
+    list_pattern_type type;
+    union {
+        list_patt_bruckt *p_brucket;
+        list_pattern_scope *p_scope;
+    } obj;
+} list_patt;
+
+typedef struct {
+    char *name;
+    pattern *p;
+} patt_field;
+
+typedef struct {
+    patt_field *pf;
+    list *p_fields;
+} record_patt;
+
+
+struct pattern {
+    pattern_type type;
+    union {
+        int_lit *i_val;
+        float_lit *f_val;
+        str_lit *s_val;
+        rstr_lit *rs_val;
+        char *ident_n;
+        list *list_pattern;
+        record_patt *record_pattern;
+    } obj;
+};
```

Comment:
why you use list for `list_pattern`?!!
just like `lit_expr` struct, use the value directly with out wrapping

```c
char *name;
int64_t i_const;
double f_const;
char *s_const;  /* no need for two type of strings */
list_patt *l_patt;
record_patt *r_patt;
```